full review post-trial 01 and new episodes in tutorials-early (#70)
epiverse-trace/tutorials-early: episodes/read-cases.Rmd

---
title: 'Read case data'
teaching: 20
exercises: 10
editor_options: 
  chunk_output_type: inline
---

:::::::::::::::::::::::::::::::::::::: questions 

- Where do you usually store your outbreak data?
- How many different data formats can I read? 

> - Is it possible to import data from database and health APIs? 

Author: do you suggest this edit?
```suggestion
- Is it possible to import data from databases and health APIs? 
```

## Reading from files 

Several packages are available for importing outbreak data stored in individual files into `R`. These include [rio](http://gesistsa.github.io/rio/), [readr](https://readr.tidyverse.org/) from the `tidyverse`, [io](https://bitbucket.org/djhshih/io/src/master/), [ImportExport](https://cran.r-project.org/web/packages/ImportExport/index.html), and [data.table](https://rdatatable.gitlab.io/data.table/). Together, these packages offer methods to read single or multiple files in a wide range of formats.

The below example shows how to import a `csv` file into `R` environment using `{rio}` package.

```{r,eval=FALSE,echo=TRUE}
library("rio")
library("here")

# read data
# e.g., the path to our file is data/raw-data/ebola_cases.csv then:

> ebola_confirmed <- read_csv(here::here("data", "raw-data", "ebola_cases.csv"))

Author: this edit is related to https://github.com/epiverse-trace/tutorials-early/issues/73. I found this facilitates running the code.

# preview data
head(ebola_confirmed, 5)
```

```{r,eval=TRUE, echo=FALSE, message=FALSE}
# read data
case_data <- rio::import(file.path("data", "ebola_cases.csv"))
# preview data
head(case_data, 5)
```

> Similarly, you can import files of other formats such as `tsv`, `xlsx`, etc.

Author: may the original one be clearer?

::::::::::::::::::::::::::::::::: challenge

###  Reading compressed data 

Take 1 minute:

> - Is it possible to read compressed data in `R`?

Author: ```suggestion

- Can you read data from a compressed file in `R`?
```

> - Is it possible to read compressed data in `R`?

Author: I suggest adding the text to download the file in this same line too.

```suggestion
Can you read data from a compressed file in `R`? Download this [zip file](https://epiverse-trace.github.io/tutorials-early/data/Marburg.zip) containing data for Marburg outbreak and then import it to your working environment.
```

::::::::::::::::: hint

You can check the [full list of supported file formats](http://gesistsa.github.io/rio/#supported-file-formats) 
in the `{rio}` package on the package website. Here is a selection of some key ones:

```{r, eval=FALSE}
rio::install_formats()
```

::::::::::::::::::::::

::::::::::::::::: solution

```{r,eval=FALSE}
rio::import(here::here("some", "where", "downto", "path", "file_name.zip"))
```

> Click [here](https://github.com/epiverse-trace/tutorials-early/tree/md-outputs-PR-39/data/Marburg.zip) to download a zip file containing data for Marburg outbreak  and then import it to your working environment.

Author: ```suggestion
Click [here](https://epiverse-trace.github.io/tutorials-early/data/Marburg.zip) to download a zip file containing data for Marburg outbreak  and then import it to your working environment.
```

::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::::::

> ## Reading from databases

Author: placed as issue in #77

The [DBI](https://dbi.r-dbi.org/) package serves as a versatile interface for interacting with database management 
systems (DBMS) across different back-ends or servers. It offers a uniform method for accessing and retrieving data from various database systems.

> 

Author: this is the source of figure https://jcsites.juniata.edu/faculty/rhodes/dbms/relnmodel.htm

> 

Author: as issue in #78

> The following code chunk demonstrates how to create a temporary SQLite database in memory, store the `case_data` as a table on it, and subsequently read it:

Author: ```suggestion
The following code chunk demonstrates how to create a temporary SQLite database in memory, store the `ebola_confirmed` as a table on it, and subsequently read it:
```

```{r,warning=FALSE,message=FALSE}
library("DBI")
library("RSQLite")

# Create a temporary SQLite database in memory
db_con <- DBI::dbConnect(RSQLite::SQLite(), ":memory:")

# Store the 'case_data' dataframe as a table named 'cases'
# in the SQLite database
DBI::dbWriteTable(db_con, "cases", case_data)
# Read data from the 'cases' table
result <- DBI::dbReadTable(db_con, "cases")
# Close the database connection
DBI::dbDisconnect(db_con)
# View the result

> base::print(utils::head(result))

Author: issue in #79
